### DISCUSSION TITLE
GPU usage maxed at 10%

### DISCUSSION THREAD

**Discussion by cowwoc** [Lightning-AI/pytorch-lightning#11580] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I was never able to get GPU usage to 10% for my own application so I decided to try BoringModel and ran into the same problem.

1. Take https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report/bug_report_model.py
2. Add `gpus=-1` to Trainer configuration
3. Increase `max_epochs` to 10_000 in Trainer configuration

Environment:

```
* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 3080
        - available:         True
        - version:           11.3
* Packages:
        - numpy:             1.21.4
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0+cu113
        - pytorch-lightning: 1.5.8
        - tqdm:              4.62.3
* System:
        - OS:                Windows
        - architecture:
                - 64bit
                - WindowsPE
        - processor:         AMD64 Family 23 Model 113 Stepping 0, AuthenticAMD
        - python:            3.9.7
        - version:           10.0.19044
```

Am I doing something wrong, or is there some sort of bug in pytorch or pytorch lightning triggering this?

**Comment by rohitgr7** [comment]
boring model is not very complex and won't lead to high GPU usage. Maybe try increasing the batch_size?

**Reply by cowwoc** [reply]
Cross-posted from https://discuss.pytorch.org/t/how-to-get-more-utilization-out-of-my-gpu/81573/9

The main problem ended up being the underlying `Dataset` but not how you’d expect. The GPU was waiting on the data from the `DataLoader` but increasing the number of workers didn’t help. Each loader was invoking `torch.randn()` to generate dummy data but it turns out this is does not generate data fast enough no matter how many workers you throw at it. Upon increasing the call with `torch.cuda.FloatTensor(size).normal_()` GPU usage shot up. I experimented with different `Dataset` lengths, `batch_size` and `num_workers`. Eventually I was able to increase GPU utilization to 50% and this is the best I could manage.

The GPU memory (10GB) was the limiting factor on the `Dataset` length and `batch_size` I could play with.

The profiler shows that the `DataLoader` is still the bottleneck, but I can’t think of any other way to improve this further. Can you?

Here is my code:

```
class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.cuda.FloatTensor(length, size).normal_()

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.batch_size = 4096
        self.feature_size = 512
        self.layer1 = torch.nn.Linear(self.feature_size, self.feature_size)
        self.layer2 = torch.nn.Linear(self.feature_size, self.feature_size)
        self.layer3 = torch.nn.Linear(self.feature_size, 2)

    def forward(self, x):
        x = self.layer1(x)
        x = self.layer2(x)
        return self.layer3(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.1)

    def optimizer_zero_grad(self, epoch, batch_idx, optimizer, optimizer_idx):
        # Set gradients to None instead of zero to improve performance
        optimizer.zero_grad(set_to_none=True)

    def train_dataloader(self) -> DataLoader:
        return DataLoader(RandomDataset(self.feature_size, self.batch_size * 256),
                          batch_size=self.batch_size,
                          pin_memory=False, num_workers=5, persistent_workers=True)

    def val_dataloader(self) -> DataLoader:
        return DataLoader(RandomDataset(self.feature_size, self.batch_size * 6),
                          batch_size=self.batch_size,
                          pin_memory=False, num_workers=2, persistent_workers=True)

    def test_dataloader(self) -> DataLoader:
        return DataLoader(RandomDataset(self.feature_size, self.feature_size * 10),
                          batch_size=self.batch_size,
                          pin_memory=False, num_workers=2, persistent_workers=True)


def run():
    model = BoringModel()
    # profiler = PyTorchProfiler(with_stack=True)
    profiler = None
    trainer = Trainer(
        gpus=GPUS,
        default_root_dir=os.getcwd(),
        num_sanity_val_steps=0,
        max_epochs=10_000,
        enable_model_summary=False,
        detect_anomaly=False,
        auto_select_gpus=True,
        profiler=profiler
    )
    trainer.fit(model)

if __name__ == "__main__":
    run()
```

**Reply by rohitgr7** [reply]
maybe try decreasing the prefetch_factor:
https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader

**Reply by cowwoc** [reply]
> maybe try decreasing the prefetch_factor: https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader

The default value is 2. You can't really decrease it much more than this :) Did you mean something else?

Also, why do you think that decreasing prefetch_factor would help? Did you mean increasing it instead?

**Reply by rohitgr7** [reply]
yeah sorry! I meant try increasing it.

but just read your full comment. If dataloader isn't able to process x data on time, wondering whether expecting it to processes Nx would even work 🤔

**Reply by rohitgr7** [reply]
wait! your data is already on the device?
```py
self.data = torch.cuda.FloatTensor(length, size).normal_()
```

**Reply by cowwoc** [reply]
> wait! your data is already on the device?
> 
> ```python
> self.data = torch.cuda.FloatTensor(length, size).normal_()
> ```

Yes. I figured this would rule out one variable.

**Comment by cowwoc** [comment]
I think I kept on playing with this and got GPU usage up to ~50%. But that's as high as I could go. As far as I can tell, GPU usage can only increase if we're computing more on the GPU before attempting to load more data, but I could not load more data onto the GPU because it would surpass the onboard memory (10GB).

You should be able to run the above testcase. Is there anything else I can do to improve performance without using more GPU ram?
